### ISSUE TITLE
Explore making read operations composable

### ISSUE THREAD

**Issue by njonsson** [njonsson/structured_io#10]
The goal is to enable complex read operations to succeed or fail atomically.

For example, it is currently impossible to consume a Tag-Length-Value (TLV) data structure without forcing the client to maintain state across read operations. This is because if the Tag is present but the Length is either absent or incomplete, the Tag must be stored for later retry because it has been consumed. What we want is to be able to perform a nested read operation that describes reading the Value as contingent upon the Length, and the Length as contingent upon the Tag, consuming nothing if the whole data structure is not present.

**Comment by njonsson** [comment]
A candidate API for this involves adding an optional function argument to each _StructuredIO.read*_ function. The client-specified function receives the result of the read operation and decides whether the outer read operation succeeded or failed.

```elixir
# This is really sketchy!
StructuredIO.read(structured_io, 2, fn tag ->
  StructuredIO.read(structured_io, 2, fn length_binary ->
    # Convert length_binary to an integer — how?
    StructuredIO.read structured_io, length
    # Return `{:ok, result}` on success or `result` on failure
  end)
  # Return `{:ok, result}` on success or `result` on failure
end)
```

**Comment by njonsson** [comment]
Perhaps all that’s needed is a _StructuredIO.transaction/2_ function that receives a PID and a function. If the provided function returns `{:ok, term}` then `term` is the result; otherwise the PID’s state is rolled back and the provided function’s return value is the result.

**Comment by njonsson** [comment]
Addressed in 0a2e94b47e868567e82623f1bd0de2006275daa4 ([*master* branch](https://github.com/njonsson/structured_io)) and d550c32c44527b745fdcc3a501065a7c84fb149c ([*v0-x* branch](https://github.com/njonsson/structured_io/tree/v0-x)).
